### PR TITLE
[RFC] [WIP] configuration via file & set different ruleset for given files

### PR DESCRIPTION
**Pull request by c33s**
i forked this nice tool because i wanted to integrate it into my ci pipeline but had problems with specific files i have no permission to change (vendor files i overwrite and changing the content of some of their variables). for example extending knpmenu can lead to code like this `{%- set childrenClasses = childrenClasses|merge({'class': 'link'}) -%}` which triggers the `LowerCaseVariable` rule.
of course i can disable the `LowerCaseVariable` completly in a custom ruleset but i just want to disable the rule for this specific file and still use all other rules for this file and use the official ruleset for all my other files. so i extended this tool, mainly with the code from phpcs fixer, to be able to do so.

new features:

- set the config in a config file `.twig_cs`. `.twig_cs.dist` and custom filename
- white/blacklsit with full finder configuration with multiple `in`, `exclude`, `path`, `notInPath`,... 
- set per file rules (can be a glob or a path)
- maybe a little bit faster because of a single finder instance

changes:
- argument name `paths` to `path`

to discuss:
- how to handle different basepaths? currently there is a bug if multiple paths are given #116 which shows the wrong basepath if multiple are given. after my change only a single finder is used which also can have multiple basepaths which i won't reccomend, as the relative path is calulated from the basepath. 
- should the magic be removed from generating the reporter service name (no auto append from `reporter.` in config. what about cli?

having the following dir structure:
```
templates
    foo
        bar.html.twig
other-templates
    foo
        bar.html.twig
```

if i provide `templates` and `other-templates` as `->in()` parameter for the finder and i have an error in both `bar.html.twig` files i will get the following path output:
```
foo/bar.html.twig
foo/bar.html.twig
```
the bug in the old behvior would lead to:
```
other-templates/foo/bar.html.twig
other-templates/foo/bar.html.twig
```
as the last path is `other-templates`. the fix would be to use multiple finder or some lookup map. i would suggest to stick with one finder with one basedir and work with `exclude()` then the relative paths are shown correct

example config `.twig_cs.dist`:

```php
<?php
$finder = FriendsOfTwig\Twigcs\Finder::create()
    ->in(__DIR__)
    ->exclude('templates')
;

return \FriendsOfTwig\Twigcs\Config\Config::create()
    ->setFinder($finder)
    ->setSeverity('warning')
    ->setReporter('reporter.console')
    ->setName('my-config')
    ->setRuleSet(FriendsOfTwig\Twigcs\Ruleset\Official::class)
    ->setSpecificRuleSets([
        '*/warning.html.twig' => FriendsOfTwig\Twigcs\Tests\Ruleset\SuppressWarningForUnusedVariable::class,
        '*/errors.html.twig' => FriendsOfTwig\Twigcs\Ruleset\Official::class,
        'tests\data\valid_corpus\form_div_layout.html.twig' => FriendsOfTwig\Twigcs\Ruleset\Official::class,
    ])
;
```

todo:
- [ ] credit authors from copied code (phpcs fixer)
- [ ] check custom config file name
- [ ] add/fix tests
- [ ] phpdoc
- [ ] check path modes
- [ ] output if finder is overridden
- [ ] fix `TODO`s
- [ ] add info to readme

### CODE REVIEW

**Comment by OwlyCode** [comment]
I really like the file based config. This is exactly what I had in mind for the future and I'm glad you made this a real thing!

Concerning the multiple base paths bug, this feels like a separate issue. I submitted a fix for the current command that uses a map (see #119). I think you should integrate it into your work so we keep the promise of multiple base paths. What about using multiple finders ?

```php
<?php
$finderA = FriendsOfTwig\Twigcs\Finder::create()->in(__DIR__.'/admin');
$finderB = FriendsOfTwig\Twigcs\Finder::create()->in(__DIR__.'/client');

return \FriendsOfTwig\Twigcs\Config\Config::create()
    ->addFinder($finderA)
    ->addFinder($finderB)
    // [...]
;
```

**Comment by c33s** [comment]
@OwlyCode i fixed the path problem with [simple-file-system][1] created by  @TomasVotruba <3

to quote his readme:

> It also fixes WTF behavior of Symfony\Component\Finder\SplFileInfo. Which one? When you run e.g. vendor/bin/ecs check src and use Finder, the getRelativeFilePath() in Symfony now returns all the relative paths to src. Which is useless, mainly with multiple dirs like: vendor/bin/ecs check src tests both containing file Post.php.

thank you very much for this awesome helper @TomasVotruba

@OwlyCode i would stick with one finder as its now fixed with using smart file system. would prefer this instead of using a map and multiple finder. is that ok for you?

[1]: https://github.com/Symplify/SmartFileSystem

**Comment by TomasVotruba** [comment]
> thank you very much for this awesome helper @TomasVotruba

Wow, I'm hapy somebody else finds it useful, thanks for reaching out :+1:

**Comment by OwlyCode** [comment]
Hey @c33s and sorry for the long time I took before looking at this. I'm not sure that introducing a new dependency to solve the finder issue is the best way to go. To me it brings new constraints for the future:

- We will have to watch out for conflicts between our dependencies, for instance when symfony/finder 6.0 comes out.
- We will introduce a risk of conflict for users with global installations of both twigcs and other libraries relying on smartfilesystem.

I recently planned to remove twig and pimple as dependencies for the same reasons as above (#114). I'm not saying we should be dependencies-free, but the benefits from the library for our use case are not that high when compared to the constraints that could come with it in the future.

I'm still convinced it's a separate issue from the file based config, and that we should focus on reducing the scope to that feature. It would be easier for me to have smaller PRs so I can review and merge them quickly than to handle big ones.

Can I do anything on my side to help on the code?

**Comment by c33s** [comment]
@OwlyCode my workload is quite high in the monent, so i had no time to carefully read your comments.

to be honest, you have taken the wind out of my sails as you requested to change the finder handling into multiple finders and to split up the PR in smaller ones.

i really thought about how to enhance the performance of twigcs (the use of https://plugins.jetbrains.com/plugin/7622-php-inspections-ea-extended-/  really helped for the code analysis) and in my opinion its a kind of all or nothing because the config file format and style is directly connected with having  one or more finder.

is it really a good idea to split it up and make a multi finder solution? 

in my opinion removing dependencies is really a bad idea. it feels like lets get back to plain php containing html, javascript and credentials only to have no dependencies.

for me tools like this (twigcs, php-cs-fixer, php_cs, phpstan, ...) are CI tools and no project dependencies. they should not (must not) be a dependency of the project i want to test. this is the reason why i use phar files (if the phar file needs external dependencies i have a 2nd composer file in a subdirectory where i can include additional rules and other helper code. helper code for the tools not for the project). 
if i use a yaml linter which is written in go or in ruby, i (my project) don't care about their dependencies, they are simply tools like `wget` with a given purpose. 
so `twig`, `pimple` and `smart-file-system` are dependencies of twigcs and it should not be thought about how to have less dependencies to not conflict with project dependencies but should be used standalone as phar file or if it is really necessary to have it as project dependency, maybe the better approach is a prefixed version (like https://github.com/rectorphp/rector-prefixed https://github.com/rectorphp/rector/tree/compiler/compiler). also twigcs is something you can install in your global composer and there you also have multiple tools to prevent conflicts (like for example https://github.com/consolidation/cgr). so why the fear of dependencies?

what do you think of my thoughts?

**Comment by OwlyCode** [comment]
Hey! @c33s I did not expect my feedback to be so demotivating. This was not the goal. So I took some days to think about all of this before answering. I see a lot of different things on different subjects here.

**Performance:** I don't know how having one or multiple finders will have an impact on performances. This should be benchmarked before any statement can be made about this. What we have now is acceptable unless we prove we can do really better. Breaking core logic or introducing a whole new dependency for a 0.1% gain might not be the choice we want to make when we get to see the numbers.

**About dependencies:** I'm not in fear about dependencies, I like when someone else does something better than me and allows me to use it. But I'm also cautious as when to introduce new dependencies because it always bring some constraints. Do we really need a whole dependency for a bug that we can easily avoid? I will not comment on the "plain php containg html" part because it is a straw man: I never said that.

**Global tool vs tool as a dependency:** I 100% agree with you that tooling should not be part of a project's dependencies, but sometimes you have no other choice: the team you work with chose differently, the CI at your workplace requires that the tool is a dev-dependency so it can be executed from the vendors and many other reasons I don't see right now. So while I agree on that point, we still have to do what is best (or the least wrong) for projects/teams that chose the other way.

You did a really great job at bringing the file based config to reality. You saved me hours of work and I'm sincerely grateful for that. If I were you I would ship it with multiple finders so it gets quickly merged, then work on a single finder version of that so we can benchmark both of the approaches and finally state on what is the best option.

**Comment by Yivan** [comment]
Hi !

@c33s Thanks a lot for this nice addition!

Just to share my experience, I am in the process of improve our CI and looking for some twig linter being opiniated (not just errors, twig:lint of symfony make it already wonderfully by checking custom tags too) to assure good formatting. Twigcs is the only tools to handle seriously this.
I try to use only phar when possible (they are in git so, each devs have the same version). And having a config file at the root project would be the must.

Actually there are just two rules which i found realy too opiniated and "old skool": no space before/after | (space bring clearly a better lecture of the code, i don't know why twig norms say to do like this... surely because rules is old one... tendance now is spacing between operators, see PSR12 for instance or prettier) and second rule is lower case variable... it is no more used in php was old times...).
So I need to disable them, but config files need composer package, etc...

Having this PR to reality would greatly help me in this use case, and many users i think too.

@OwlyCode If not merged, at last being able to disable rules from command line would be a reallity great addition. But sure having config file is the best so developper have just to configure twigcs in their IDE wihout having adding extra parameters. For post hooks it is less problematic as we can add parameters to commands.

Best regards.

**Comment by OwlyCode** [comment]
Hi @Yivan! Disabling/customizing rules on-demand or per project has been a long running topic about twigcs. I agree that this PR should be merged as soon as possible because it would be a key feature. I know that Julian has been quite busy lately (and so am I sadly).

@c33s would you mind if I finished this PR?

**Comment by Yivan** [comment]
@OwlyCode Thanks for your input on this.

For now a shell wrapper is used, but this is ugly : ) But it works waiting for this feature ; )

For people interested, it could help:

```
for ELT in "${SRC_PATH[@]}"; do
    CURRENT_OUTPUT="$(php -d short_open_tag=1 ./tests/phar/twigcs.phar --severity error "$ELT")"
    OUTPUT="$(echo -e "${OUTPUT}\\n${CURRENT_OUTPUT}")"
done
OUTPUT="$(echo "$OUTPUT" | (! grep -v 'WARNING ' ) \
    | (! grep -v '0 space before the "|"' ) \
    | (! grep -v '0 space after the "|"' ) \
    | (! grep -v 'variable should be in lower case' ) \
    | sed '/\.twig/{$!N;/\n.*\.twig/!P;D}' | sed '/\.twig/{$!N;/\n.* violation(s) found/!P;D}' | (! grep -v 'violation(s) found' ) | sed '/^$/d')"
set -e
if [ -n "$OUTPUT" ]; then echo -e "\\n$OUTPUT"; echo -e "\\nError: some file(s) are not formatted."; exit 1; fi
```

As you see I added too the possibility to pass multiple files and agregate output.
Because we use husky + lint-staged which give in parameters several file full path, and twigcs only acept one file. Maybe it could be another improvement to support something like:
`twigcs.phar /path/file1.twig /path/to/a/folder /other/path/files2` ; )

**Comment by k00ni** [comment]
Hi, can anyone give me a short status update, please? Thanks.

File: src/Console/LintCommand.php [hunk 1]
```diff
@@ -2,96 +2,100 @@
...
+//                $file->getRelativeFilePath(),
+//                $file->getRelativePathname(),
+//                ''
+//                'getRelativeFilePathFromCwd',
+//                $file->getRelativeFilePathFromCwd(),
+//                'getRelativePathname',
+//                $file->getRelativePathname(),
+//                'getRelativeFilePathFromDirectory',
+//                $file->getRelativeFilePathFromDirectory(getcwd())
+//            );
```
Comment (k00ni) [inline src/Console/LintCommand.php:75]: Code which is commented out shouldn't be part of the PR. If it is required for refactoring one should restore it using the version control system at hand.

File: src/Console/LintCommand.php [hunk 2]
```diff
@@ -2,96 +2,100 @@
...
-        foreach ($files as $file) {
-            $violations = array_merge($violations, $container['validator']->validate(new $ruleset(), $container['twig']->tokenize(new \Twig\Source(
+            $ruleset = $resolver->getRuleset($file->getRelativePathname());
+            $currentViolations = $validator->validate($ruleset, $twig->tokenize(new \Twig\Source(
                 file_get_contents($file->getRealPath()),
                 $file->getRealPath(),
-                str_replace(realpath($path), rtrim($path, '/'), $file->getRealPath())
-            ))));
+                // TODO: relative path is without base dir if multiple in's are given.
+//                $file->getRelativePathname(),
```
Comment (k00ni) [inline src/Console/LintCommand.php:83]: Commented out code again (see comment above). Also broken coding style in my opinion.

File: src/Config/ConfigurationResolver.php [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+        'config' => null,
+    ];
+
+    private $finder;
+
+    /**
+     * @var Container
+     */
+    private $container;
+    /**
```
Comment (k00ni) [inline src/Config/ConfigurationResolver.php:57]: Just a tiny thing: maybe add a line above between the comment and the variable?

File: src/Config/ConfigurationResolver.php [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+        if (null === $this->specificRulesets) {
+            $this->specificRulesets = $this->getConfig()->getSpecificRulesets();
+        }
+
+        $file = $this->normalizePath($file);
+        foreach ($this->specificRulesets as $pattern => $rulesetClassName) {
+            // TODO: fnmatch allow setting flags https://www.php.net/manual/en/function.fnmatch.php
+            $pattern = $this->normalizePath($pattern);
+            if ($file === $pattern || fnmatch($pattern, $file)) {
+
```
Comment (k00ni) [inline src/Config/ConfigurationResolver.php:145]: That extra line should be removed in my opinion.

File: src/Config/ConfigurationResolver.php [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+            default:
+                throw new \InvalidArgumentException('Invalid severity limit provided.');
+        }
+    }
+
+    public function getConfig(): ConfigInterface
+    {
+        if (null === $this->config) {
+            foreach ($this->computeConfigFiles() as $configFile) {
+                if (!file_exists($configFile)) {
```
Comment (k00ni) [inline src/Config/ConfigurationResolver.php:187]: Because some functions are prefixed with a `\`, why not `file_exists` too?

**Review by k00ni — COMMENTED** [review]
Good work! I hope my comments help a bit.

File: src/Console/LintCommand.php (hunk 1 above)
Comment (c33s) [reply src/Console/LintCommand.php:75]: thats of course clear. it's a RFC PR

File: src/Config/ConfigurationResolver.php (hunk 3 above)
Comment (c33s) [reply src/Config/ConfigurationResolver.php:57]: well would be good to define a coding standard that can be automatically applied (with php-cs-fixer or ecs). not knowing what the coding standard is i prefere to have those properties together

File: src/Config/ConfigurationResolver.php (hunk 4 above)
Comment (c33s) [reply src/Config/ConfigurationResolver.php:145]: same as for the other code style comment

**Comment by c33s** [comment]
@k00ni
 
> Hi, can anyone give me a short status update, please? Thanks.

well the status is canceled. as @OwlyCode doesn't like my required dependencies and his requested changes are conflicting with my workload and my point of view about dependencies i close this PR.

**Comment by k00ni** [comment]
Thanks for commenting, sad to see it being closed.

**Comment by c33s** [comment]
@k00ni if you don't need the newest features from here (multiple twig version support) you can use my fork as i do. 

```
{
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/c33s/twigcs.git"
        }
    ],
    "require": {
        "friendsoftwig/twigcs": "dev-feature/config",
    }
}
```

if i upgrade to twig 3 i a sure i will have again a look at this PR. hope then i will find a solution which will allow my code to be merged.
